capstone/src/main/webapp/sheetsScript.js

```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
+  let totalText = '';
+  let range = '!A1:YY';
+  for (const name of sheetNames) {
+    const sheet = await getSpreadsheet(id, name, range);
+    const text = await getTextFromSheet(sheet);
+    totalText += text + '\n';
+  }
+  handleInput(totalText, langElement.value, requestedAttributes, tokenizer);
+
+  // Create spreadsheet if requested
+  const userDecisionElement = document.getElementById('sheets-output-yes-no');
+  if (userDecisionElement != null && userDecisionElement.value === 'yes') {
+    let title = 'Perspective Output';
+    const titleElement = document.getElementById('sheets-title');
+    if (titleElement != null && titleElement.value != '') {
+      title = titleElement.value;
+    }
+
+    const newSheetId = await createSheet(title);
+
+    // Check if user has entered a custom range
+    const userRangeElement = document.getElementById('sheets-range-yes-no');
+    if (userRangeElement != null && userRangeElement.value === 'yes') {
+      // Check if this is a valid Sheets range
+      const rangeElement = document.getElementById('sheets-range');
+      if (rangeElement != null) {
+        const rangeRegEx = rangeElement.value.match(/(.+)(![\w]*[\d]*(:[\w]*[\d]*)?)/);
+        if (rangeRegEx != null) {
+          sheetNames = [rangeRegEx[1]];
+          range = rangeRegEx[2];
+        } else {
+          alert('Please input a range in a valid format.');
+        }
+      }
+    }
+
+    // Make the new spreadsheet mirror sheet names of the input
+    if (sheetNames[0] != 'Sheet1') {
+      await preprocessSheet(newSheetId, sheetNames);
+    }
+
+    for (let i = 0; i < sheetNames.length; i++) {
+      const body = await createSheetOutput(id, sheetNames[i], range, langElement.value, requestedAttributes);
+      await appendDataToSheet(newSheetId, sheetNames[i], body);
+      const numRows = body.length;
+      const numCols = body[0].length;
+      const sheetId = await getSheetId(newSheetId, sheetNames[i]);
+      await addFormatting(newSheetId, sheetId, numRows, numCols);
+    }
+  }
 }
 
 /**
- * Returns all text in a Google Sheet with an id of spreadsheetId
+ * Returns the name(s) of every sheet in the Google Sheet with an id of id
  */
-async function getTextFromSheet(spreadsheetId) {
+async function getSheetNames(id) {
+  const response = await gapi.client.sheets.spreadsheets.get({
+    spreadsheetId: id,
+  });
+  const sheets = await response.result.sheets;
+  const sheetNames = [];
+  for (const sheet of sheets) {
+    sheetNames.push(sheet.properties.title);
+  }
+  return sheetNames;
+}
+
+/**
+ * Returns the id of a sheet in the Google Sheet
+ * with corresponding spreadsheet id and name
+ * and returns 0 if not found
+ */
+async function getSheetId(id, name) {
+  const response = await gapi.client.sheets.spreadsheets.get({
+    spreadsheetId: id,
+  });
+  const sheets = await response.result.sheets;
+  for (const sheet of sheets) {
+    if (sheet.properties.title === name)
+    return sheet.properties.sheetId;
```

Comment:
nit: Should there be an extra indent here?

Reply:
Yes - good catch